README
Expect-Angel version 1.00
=========================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Expect-1.21
  Term::ReadKey

COPYRIGHT AND LICENCE

These modules can be used under the same terms as Perl.

Copyright (C) 2010 by Ming Zhang

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.0 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Expect::Angel.

0.01  Thu Apr 22 15:30:37 2007
	- original version; created by h2xs 1.23 with options
		-AXn Expect::Angel

0.02  Thu Mar 05 15:10:07 2008
        - add some doc


1.00  Sun Oct 10 2010
        - major change in both methods and implementation
        - retired echo(), replaced by cmdexe()
        - a lot other features to support variety of types of devices including Cisco, Juno, and Linux



t/Linux.pm
package Linux;
#
# This is example shows complex DUT scenarios by means of Linux machine
# 
# there are 5 states
# 0 - initial state
# 1 - prompt: [~/root]$
# 2 - prompt: [~/root/s1]$
# 3 - prompt: [~/root/s1/s2]$
# 4 - prompt: [~/root/s1/s2/s21]$
# 5 - prompt: [~/root/s1/s2/s22]$
#
# 0 --> 1 --> 2 --> 3 --> 4
#                   |---> 5

# state transition
# by "cd" command
# 
use Expect::Angel;
@ISA = ("Expect::Angel");
my $debug = 0;
#
sub new {
   my $type = shift;
   my $para = {@_};
   my $conn = $type->build(%$para);

   #
   # define states
   #
   my $states = [( { prompt => '\[~/root\]\$ $' },
                   { prompt => '\[~/root/s1\]\$ $' },
                   { prompt => '\[~/root/s1/s2\]\$ $' },
                   { prompt => '\[~/root/s1/s2/s21\]\$ $' },
                   { prompt => '\[~/root/s1/s2/s22\]\$ $' },
                )];
   $conn->addState(%$_) for (@$states);

   #
   # define transition
   #
   my $trans = {command => "telnet $para->{dut}",
                expect => { 'login: \r?$' => $para->{user},
                            'Password: \r?$' => $para->{passwd} }
               };
   $conn->transitState(0,1,$trans);
   $trans = {command => "cd s1", expect => { } };
   $conn->transitState(1,2,$trans);
   $trans = {command => "cd s2", expect => { } };
   $conn->transitState(2,3,$trans);
   $trans = {command => "cd s21", expect => { } };
   $conn->transitState(3,4,$trans);
   $trans = {command => "cd s22", expect => { } };
   $conn->transitState(3,5,$trans);
   $trans = {command => "cd ..", expect => { } };
   $conn->transitState(5,3,$trans);
   $conn->transitState(4,3,$trans);
   $conn->transitState(3,2,$trans);
   $conn->transitState(2,1,$trans);
   $trans = {nexthop => 3 };
   $conn->transitState(5,4,$trans);
   $conn->transitState(4,5,$trans);
   $trans = {command => "exit", expect => { } };
   $conn->transitState(1,0,$trans);
   if (exists $para->{'goto'}) {
      return $conn->movState($para->{'goto'}) ? $conn : undef;
   }
   $conn;
}


1;



t/testLinux.pl
#!/usr/bin/perl -w
use Expect::Angel;
use lib(".");
use Linux;
my $m = "localhost";

my $fn = 'mylog';
our $dut = new Linux(debug => 0, user => 'angel', passwd => 'qatest', dut => $m, goto => 0 , log => $fn) or die "Failed to access $s1 mode\n";

for $s1 ( 1 ... 5 ) {
   $dut->movState($s1) or die "\nfailed to back from $s2 to state $s1\n\n";
   for $s2 (1,2,3,4,5) {
      if ($dut->movState($s2)) {
         print "success from $s1 to $s2\n";
         for ($dut->echo("pwd")) {
            print "$_";
         }
      }else{
         print "failed from $s1 to $s2\n";
         print "current state: $dut->{current}\n";
      }
   }
}


lib/Expect/Angel/Cisco.pm
package Expect::Angel::Cisco;
use base Expect::Angel;
our $VERSION = '1.00';

=head1 NAME

 Expect::Angel::Cisco - Build up a robust connection to Cisco or cisco-like device


=head1 SYNOPSIS

 use Expect::Angel::Cisco;
 my $dut_ip = '10.1.1.100';
 # connects the device and goes to its enable mode
 my $dut = new Expect::Angel::Cisco(connection => "ssh $dut_ip", goto => 'enable' )
             or die "Failed to access to device\n";
 # check device's clock
 my $clock = $dut->cmdexe("show clock");

 # configure its interface eth1/1
 my @int_cmds = ("int eth1/1",
                 "ip address 10.2.1.1/24",
                 "no shut"
                );
 $dut->cmdexe(\@int_cmds, "config");

 # parse its parameters on eth1/1
 my @lines = $dut->cmdexe("show int eth1/1");
 for (@lines) {
     # parsing
 }

 # executes a list of commands at "config" mode of Cisco device
 my @cmd = ("interface e0/0",
            "ip address 10.1.1.10 255.2555.255.0",
            "no shut",
            "exit",  
            "interface e0/1",
            "ip address 10.1.2.100 255.2555.255.0",
            "no shut"
           );
 $dut->cmdexe(\@cmd, "config");

 # always stay at config mode for each command execution, although
 # existing some command that may change mode inadvertently.
 $dut->sticky(1);
 my @cmd = ("interface e0/0",
            "ip address 10.1.1.10 255.2555.255.0",
            "end",  # NOTE, this cause the transition to "enable" mode
            "interface e0/1",
            "ip address 10.1.2.100 255.2555.255.0",
            "no shut"
           );
 $dut->cmdexe(\@cmd, "config");
  

 # check if a command is rejected by device
 if ( defined (my $clock = $dut->cmdSendCheck("show clock")) ) {
     print "system current clock is \"$clock\"\n";
 }else{
     # "show clock" is rejected 
 }


=head1 Description

 This module is a derivation of Expect::Angel on Cisco-like devices. It's pretty simple 
because most of the function has been implemented in Expect::Angel. It serves more like
an example of how to use Expect::Angel to apply to Cisco device.

=head1 new()

  Create a new Cisco object 
  connection => "command to connect to Cisco device", for example telnet 1.1.1.1 3000
  hostname => hostname_of_Cisco, it should be switch's host name. It will be used in prompt 
              pattern match. If not specified, it's ignored at prompt match, but risky.

  # user/passwd/enablePass, if not provided, will be prompted to input from keyboard
  user     => username, used to access the device.
  passwd   => password, used to access the device.
  enablePass  => password, used to access the enable mode.
  goto => enable|config, optional. Will go to that state (mode) if specified.
  see Angel's build method for all other options.

  This is a good example to show how to build a module by means of Expect::Angel.
  Different type of devices may have slight difference in their state transition.
  You can modify it accordingly to fit your specific device. The major methods are
  inherited from Expect::Angel.
  
=cut

sub new {
     my $type = shift;
     my $para = {@_};

     # default pattern is always added into Expect body. 
     my $defPats = [{ '[-]+\s*(More|more)\s*[-]+' => ' ',
                     '^.* memory\? *\[confirm\] *$' => '\r',
                     '^.*Are you sure\?.*\[confirm\] *$' => '\r',
                     '^.*Are you sure\?.*$'              => 'yes\r',
                     '^.*Are you sure you want to continue connecting\?.*$' => 'yes\r',
                     'Escape character is \\\'\^\]\\\'\.' => '\r',
                   }];

     $para->{defPattern} = $defPats;
     $para->{aggressive} = 1;
     my $conn = $type->build(%$para);
     
     # hostname is used at prompt match, it should be switch's host name.
     # if not specified, it's ignored at prompt match, but risky.
     my $hostname = (exists $para->{hostname}) ? $para->{hostname} : '';

     #
     # define states
     #   1 - rommon
     #   2 - non privilege
     #   3 - enable
     #   4 - config
     my $states = [( 
                   { prompt => "rommon> ?\$",       # state 1
                     name   => "rommon",         
                     descr  => "rom monitor"
                   },
                   { prompt => "$hostname> ?\$",    # state 2
                     name   => "non_privilege",         
                     descr  => "non_privilege mode"
                   },
                   { prompt => "$hostname# ?\$",    # state 3
                     name   => "enable",         
                     descr  => "enable mode"
                   },
                   { prompt => "$hostname\\\(conf(ig)?.*\\\)# ?\$",   # state 4
                     name   => "config",
                     descr  => "configuration model"
                   },
                )];
     $conn->addState(%$_) for (@$states);

     #
     # check if credential is provided, otherwise read from keyboard
     #
     $para->{user} = $conn->getUsername() unless(defined $para->{user});
     $para->{passwd} = $conn->getPasswd("Password:") unless(defined $para->{passwd});
     $para->{enablePass} = $conn->getPasswd("enable password:") unless(defined $para->{enablePass});

     #
     # define transition
     #

     # 0 -> 1 
     $trans = {nexthop => 2};
     $conn->transitState(0,1,$trans);

     # 0 -> 2, 0 -> 3 
     # right after telnet or ssh to a device, it's at 2 or 3
     my $trans = { command => $para->{connection},
                   expect => [{ 'Username: \r?$' => "$para->{user}",
                                  'login: \r?$' => "$para->{user}",
                                  'Login: \r?$' => "$para->{user}",
                               'Password: \r?$' => "$para->{passwd}",
                               'password: \r?$' => "$para->{passwd}"}]
                 };
     $conn->transitState(0,2,$trans);
     $conn->transitState(0,3,$trans);

     # 1 -> 2
     # at rommon mode, boot can go to 2.
     $trans = {command => "boot"};
     $conn->transitState(1,2,$trans);

     # 2 -> 3 
     $trans = { command => "enable",
                expect => [{ 'Password: \r?$' => "$para->{enablePass}"}]
              };
     $conn->transitState(2,3,$trans);

     # 3 -> 4 
     $trans = {command => "config t" };
     $conn->transitState(3,4,$trans);

     # 4 -> 3 
     #$trans = {command => "end", expect => { } };
     $trans = {command => "end" };
     $conn->transitState(4,3,$trans);

     # 3 -> 2
     $trans = {command => "exit" };
     $conn->transitState(3,2,$trans);

     #3 -> 0, 2 -> 0
     $trans = {command => "exit",
               expect => [{ 'telnet> \r?$' => 'quit\n' },
                          {'.+' => '\x1d' }
                          #{'.+' => '\035' }
                         ]
              };
     $conn->transitState(3,0,$trans);
     $conn->transitState(2,0,$trans);

     # 
     # transit to the specified state
     #
     if (exists $para->{'goto'}) {
        return $conn->movState($para->{'goto'}) ? $conn : undef;
     }
     $conn;
}


=head1 enableMode()

  transits to enable state. It will try 3 times (defined by errTries)
  input: none
  output: true - the state is transitted to enable state
          undef - can not transit to enable state

=cut

sub enableMode {
     my ($conn) = shift;
     $conn->movState("enable") or return undef;
}
  
=head1 configMode()

  transits to config state. It will try 3 times (defined by errTries)
  input: none
  output: true - the state is transitted to config state
          undef - can not transit to config state

=cut

sub configMode {
     my ($conn) = shift;
     $conn->movState("config") or return undef;
}

=head1 cmdexe(cmd,expect,state)

   Inherited from Expect::Angel, refer to it for detail

=cut

=head1 cmdSendCheck($cmd,$errorPattern)

 This sub will exectues $cmd at current state and detect if $errorPattern 
 matches its output in the each execution. 
 If the command is rejected, it will return immediatedly.
 $cmd: single or ref to a list of commands
 $errorPattern: optional, RE of error pattern, default is '^%|Invalid command'
 return: undef when $errorPattern matches, otherwise output of the last command

=cut

sub cmdSendCheck {
    my ($self, $cmd,$errorPattern) = @_;
    $errorPattern ||= '^%|Invalid command';
    $self->SUPER::cmdSendCheck($cmd,$errorPattern);
}


=head1 bye()

  Close the connection decently

=cut

sub bye {
    my ($conn) = shift;
    $conn->movState(0) or return undef;
}
  
=head1 pingable($ip_addr) 

 Test if target $ip_addr is reachable from the DUT by ping it.
 Input: $ip_addr - ip address of target 
 Return: number of echoback packats, all parameters are device default.
        so it will be 5 if 100% success.

=cut

#  Sending 5, 100-byte ICMP Echos to 10.129.144.1, timeout is 2 seconds:
## !!!!!
#  Success rate is 100.0 percent (5/5), round-trip min/avg/max = 0/3/16 (ms)
#  Sending 5, 100-byte ICMP Echos to 10.10.101.111, timeout is 2 seconds:
#  .....
#  Success rate is  0.0 percent (0/5)

sub pingable {
    my ($conn,$ip_addr) = @_;
    my $success;
    my $expBody = [{
                    'Target IP address or host   : ' => "$ip_addr",
                    'Repeat Count \[5\]    : ?'     => '', 
                    'Datagram size \[100\] : ?'     => '',
                    'Timeout in secs \[2\] : ?'     => '',
                    'Extended commands \[n\] : ?'   => '',
                    'Sweep range of sizes \[n\]: ?' => '',
                  }];
    for ( $conn->cmdexe("ping",$expBody,"enable") ) {
        next unless ( /^\!|\.$/ );
        $success = grep { $_ eq '!' } split '', $_;
    }
    return $success;
}

1;



lib/Expect/Angel/Juniper.pm
package Expect::Angel::Juniper;
use base Expect::Angel;
our $VERSION = '1.00';

=head1 NAME

Expect::Angel::Juniper - Build up a robust connection to Juniper device

=head1 SYNOPSIS

 use Expect::Angel::Juniper;
 my $connectCmd = "ssh 10.1.1.1";
 my $dut = new Expect::Angel::Juniper(connection => "$connectCmd", goto => 'operation' )
             or die "Failed to access to device by running '$connectCmd'\n";


=head1 Description

 This module is a derivation of Expect::Angel on Juniper Juno devices. It's pretty simple 
because most of the function has been implemented in Expect::Angel. It serves more like
an example of how to use Expect::Angel to apply to Juno OS device.

=head2 new()

  Create a new Juniper object 
  connection => "command to connect to Juniper device", for example telnet 1.1.1.1 3000
  hostname => hostname_of_Juniper, it should be device's host name. It will be used in prompt 
              pattern match. If not specified, it's ignored at prompt match.

  # user/passwd/configPass, if not provided, will be prompted to input from keyboard
  user     => username, used to access the device.
  passwd   => password, used to access the device.
  configPass  => password, used to access the configuration mode.

  goto => operation|config, optional. Will go to that state (mode) if specified.

  see Angel's build method for all other options.

=cut

sub new {
     my $type = shift;
     my $para = {@_};
     my $defPats = [{ '[-]+\s*\((More|more).*\)\s*[-]+' => ' ',
                     '^.* memory\? *\[confirm\] *$' => '\r',
                     '^.*Are you sure\?.*\[confirm\] *$' => '\r',
                     '^.*Are you sure\?.*$'              => 'yes\r',
                     '^.*Are you sure you want to continue connecting\?.*$' => 'yes\r',
                   }];

     $para->{defPattern} = $defPats;
     $para->{aggressive} = 1;
     $para->{autofillup} = 1;  # command echoed fully back behavior.
     my $conn = $type->build(%$para);
     
     # hostname is used at prompt match, it should be device's host name.
     # if not not specified, any would match
     my $hostname = (exists $para->{hostname}) ? uc $para->{hostname} : '.+';

     #
     # check if credential is provided, otherwise read from keyboard
     #
     #$para->{sshphrase} = $conn->getPasswd("SSH phrase:") unless(defined $para->{sshphrase});
     $para->{user} = $conn->getUsername() unless(defined $para->{user});
     $para->{passwd} = $conn->getPasswd("Password:") unless(defined $para->{passwd});
     $para->{configPass} = $conn->getPasswd("config password:") unless(defined $para->{configPass});

     #
     # user@host
     #
     my $prompt_kw = "$para->{user}\@$hostname";
     #
     # define states
     #   1 - operation
     #   2 - configure
     my $states = [( 
                   { prompt => "$prompt_kw> ?\$",    # state 1
                     name   => "operation",         
                     descr  => "operational mode"
                   },
                   { prompt => "$prompt_kw# ?\$",    # state 2
                     name   => "config",         
                     descr  => "configuration mode"
                   },
                )];
     $conn->addState(%$_) for (@$states);

     #
     # define transition
     #

     # 0 -> 1, 0 -> 2
     # connection => "ssh tftp1.ops.iso.test.sp2 ssh -c \"ulimit -H -t 10 ; telnet 10.129.160.247\"",
     my $trans = { command => $para->{connection},
                   expect => [{ 'Username: \r?$' => "$para->{user}",
                                  'login: \r?$' => "$para->{user}",
                                  'Login: \r?$' => "$para->{user}",
                               'Password: \r?$' => "$para->{passwd}",
                               'password: \r?$' => "$para->{passwd}"}]
                 };
     $conn->transitState(0,1,$trans);
     $conn->transitState(0,2,$trans);

     # 1 -> 2 
     $trans = { command => "config",
                expect => [{ 'Password: \r?$' => "$para->{configPass}"}]
              };
     $conn->transitState(1,2,$trans);

     # 2 -> 1
     #$trans = {command => [qw(top exit)], expect => { } };
     $trans = {command => [qw(top exit)] };
     $conn->transitState(2,1,$trans);

     # 1 -> 0
     #$trans = {command => "exit", expect => { } };
     $trans = {command => "exit" };
     $conn->transitState(1,0,$trans);

     # 
     # states are defined, now check it.
     #

     # 
     # transit to the specified state
     #
     if (exists $para->{'goto'}) {
        return $conn->movState($para->{'goto'}) ? $conn : undef;
     }
     $conn;
}

=head2 operMode()

  Transits to operational mode. It will try 3 times (defined by errTries)
  input: none
  output: true - the state is transitted to operational mode
          undef - can not transit to operational mode
=cut

sub operMode {
     my ($conn) = shift;
     $conn->movState("operation") or return undef;
}
  
=head2 configMode()

  Transits to config state. It will try 3 times (defined by errTries)
  input: none
  output: true - the state is transitted to config state
          undef - can not transit to config state
=cut

sub configMode {
     my ($conn) = shift;
     $conn->movState("config") or return undef;
}

=head2 bye()

  Close the connection decently

=cut

sub bye {
     my ($conn) = shift;
     $conn->movState(0) or return undef;
}
  
1;



lib/Expect/Angel/Linux.pm
package Expect::Angel::Linux;
use base Expect::Angel;
our $VERSION = '1.00';

=head1 NAME

Expect::Angel::Linux - Build up a robust connection to Lunix/Unix host

=head1 SYNOPSIS

 use Expect::Angel::Linux;
 my $dut = new Expect::Angel::Linux(connection => "ssh $host", liveExp => 1)
             or die "Failed to access to device by running '$connectCmd'\n";


=head2 new()

  Create a new object 
  connection => "command to connect to the host", for example telnet 1.1.1.1 3000

  echodetect => 1, see login() why you might want to set it.
  see Angel's build method for all other options.

=cut

sub new {
     my $type = shift;
     my $para = {@_};
     my $defPats = [{ '^.*Are you sure\?.*\[confirm\] *$' => '\r',
                     '^.*Are you sure\?.*$'              => 'yes\r',
                     '^.*Are you sure you want to continue connecting\??.*$' => 'yes\r',
                   }];

     $para->{defPattern} = $defPats;
     $para->{aggressive} = 1;

     # unless echodetect is set
     $para->{noechoback} = (exists $para->{echodetect} && $para->{echodetect}) ? 0 : 1;

     chomp ($para->{connection} = `which bash`) if ( $para->{connection} eq 'localhost' );
     my $conn = $type->build(%$para);
     
     #
     # define states
     #   1 - shell
     my $states = [( 
                   { prompt => '[#$%>] ?$',       # state 1
                     name   => "login",         
                     descr  => "shell on remote"
                   },
                )];
     $conn->addState(%$_) for (@$states);

     #
     # check if credential is provided, otherwise read from keyboard
     # user may be avaiable from system call getpwuid
     #
     (defined $para->{user}) or $para->{user} = getpwuid($>) or $para->{user} = $conn->getUsername();
     $para->{passwd} = $conn->getPasswd("Password:") unless(defined $para->{passwd});
     $para->{sshphrase} = $para->{passwd};
     # add them as the properties of the object
     $conn->{user} = $para->{user};
     $conn->{passwd} = $para->{passwd};

     my $authExpect = { 'Username: ?\r?$' => "$para->{user}",
                            'login: ?\r?$' => "$para->{user}",
                            'Login: ?\r?$' => "$para->{user}",
                         'Password: ?\r?$' => "$para->{passwd}",
                         'sshphrase:\r?$' => "$para->{sshphrase}",
                         'password: ?\r?$' => "$para->{passwd}"
                      };

     push @{$conn->{defPattern}}, $authExpect;

     #
     # define transition
     #

     # 0 -> 1
     # connection => "ssh tftp1.ops.iso.test.sp2 ssh -c \"ulimit -H -t 10 ; telnet 10.129.160.247\"",
     my $trans = { command => $para->{connection},
                   expect => [($authExpect)]
                 };
     $conn->transitState(0,1,$trans);

     # 1 -> 0,
     $trans = {command => "exit" };
     $conn->transitState(1,0,$trans);


     # 
     # transit to the specified state
     #
     if (exists $para->{'login'} and $para->{'login'}) {
         $conn->login() or return undef;
     }
     $conn;
}

=head2 login()

Executes the connection command and expect the shell prompt. If this is done,
it tries if the echo back is supported when raw terminial is set. Linux is not
like Cisco/Juniper routers that echo back the commands. Linux host may depend
on some configuration. If you are not sure about this, you can enable the 
auto-detect by set "echodetect => 1" in new() call. User may feel a delay in 
this auto-detection process.

input: none
output: true - success
          undef - fail
=cut

sub login {
    my ($conn) = shift;
    $conn->movState("login") or return undef;
    $conn->{state}[1]{prompt} = 'MYCOOLCONNECTION> ?$';

    if (exists $conn->{echodetect} && $conn->{echodetect}) {
        # Not sure ssh does echo command like Cisco/Juniper device
        # but raw is still highly demanded, don't do $para->{expRaw} = 0;
        # let's detect it
        my $echoTimeout = 10;
        print STDERR "\ndetecting echo back for $echoTimeout seconds, patient ...\n";
        eval { $conn->sendCmd("CANNEVERBEALEGALCOMMANDHAHABLAR", $echoTimeout, 1) };
        $conn->{noechoback} = ($@) ? 1 : 0;
    }
    $conn->cmdexe('PS1="MYCOOLCONNECTION> "');
    return 1;
}


=head2 cmdexe($cmd,$expect)

   Run a command(s), see Expect::Angel for detail

=head3 Usage
   my $output = $host->cmdexe("cat ~/myfile");
   my @output = $host->cmdexe("cat ~/myfile");

=head3 Inputs

=over

=item $cmd

 The shell command to execute on remote host

=item $expect

 Optional, ref to list of hash that describes the interactive expect body for this command.

=back

=head3 Returns

 the output of the last command execution.
 In list context it returns line by line, without \n, 
 in scalar context, it returns the string of the output.

=cut

sub cmdexe  {
    my ($self, $cmd, $expect) = @_;
    # escape single quote if any
    $cmd =~ s/'/'\\''/g;
    $self->SUPER::cmdexe($cmd,$expect);
}

=head2 cmdexe_nonesc($cmd,$expect)

Same as cmdexe() except that it does not escape single quote.
If you have a sngle quote in $comd, you mean it's used in shell command format, so it's not escaped.

=cut

sub cmdexe_nonesc  {
    my ($self, $cmd, $expect) = @_;
    $self->SUPER::cmdexe($cmd,$expect);
}

=head2 cmdSendCheck()

   Run a command and check its exit code by "echo $?".

=head3 Usage
   my ($exitcode,$output) = $host->cmdSendCheck("cat ~/myfile");

=head3 Inputs

=over

=item $cmd

 The shell command to execute on remote host

=item $expect

 Optional, ref to list of hash that describes the interactive expect body for this command.

=back

=head3 Returns

  $exitcode,$output
  if $cmd execution failed, both are undef

=cut

sub cmdSendCheck {
    my ($self, $cmd, $expect) = @_;
    my ($msg,$exitCode) = (undef, undef);
    $check = 'echo $?';
    $msg = $self->cmdexe($cmd,$expect);
    chomp ( $exitCode = $self->cmdexe($check) ) if (defined $msg);
    #chomp $exitCode;
    return ($exitCode,$msg);
}

  
=head2 existsDir($dir)

Check if $dir exists
input: $dir - name of the directory to be checked
output: 1|0 for exists or non-exists

=cut

sub existsDir {
    my ($self, $dir) = @_;
    my $output = $self->cmdexe("ls -ld $dir | cat ");
    return ($output =~ /^d/) ? 1 : 0;
}

=head2 os()

OS of the host
return: Linux|FreeBSD

=cut
sub os {
    my ($conn) = shift;
    chomp(my $os = $conn->cmdexe("uname -s"));
    return $os;
}

=head2 bye()

  Close the connection decently

=cut

sub bye {
     my ($conn) = shift;
     $conn->movState(0) or return undef;
}
  
1;

